content/docs/rfcs/32/README.md

```diff
@@ -0,0 +1,457 @@
+for the given epoch. 
+The data can be deleted when the epoch passes. 
+Storing metadata is required, so that if a user sends more than the 2 messages per epoch, 
+they can be slashed and removed from the protocol. 
+The metadata stored contains the `x`, `y` shares and the `nullifier` for the user for each message. 
+If enough such shares are present, the user's secret can be retreived.
+
+One way of storing received metadata (`messaging_metadata`) is the following format: 
+
+```
+{
+    [epoch]: {
+        [nullifier]: {
+            x_shares: [],
+            y_shares: []
+        }
+    }
+}
+```
+
+### Verification
+
+The output message verification consists of the following steps:
+- `epoch` correctness
+- non-duplicate message check
+- `zk_proof` verification
+- spam verification
+
+**1. `epoch` correctness**
+Upon received `output_message`, first the `epoch` field is checked, 
+to ensure that the message matches the current epoch. 
+If the epoch is correct the verification continues, otherwise the message is discarded.
+
+**2. non-duplicate message check**
+The received message is checked to ensure it is not duplicate. 
+The duplicate message check is performed by verifying that the `x` and `y` fields do not exist in the `messaging_metadata` object. 
+If the `x` and `y` fields exist in the `x_shares` and `y_shares` array for the `epoch` and the `nullifier` the message can be considered as a duplicate. 
+Duplicate messages are discarded.
+
+**3. `zk_proof` verification**
+
+The `zk_proof` should be verified by providing the `zk_proof` field to the circuit verifier along with the `public_signal`:
+
+```
+[
+    y,
+    merkle_proof.root,
+    nullifier,
+    signal_hash,
+    epoch,
+    rln_identifier
+]
+```
+
+If the proof verification is correct, 
+the verification continues, otherwise the message is discarded.
+
+**4. spam verification**
+
+After the proof is verified The `x`, and `y` fields are added to the `x_shares` and `y_shares` arrays of the `messaging_metadata` `epoch` and `nullifier` object. 
+If the length of the arrays is equal to the spam threshold (`limit`), the user can be slashed.
+
+### Slashing
+
+After the verification, the user can be slashed if enough shares are present to reconstruct their `identity_secret_hash` from `x_shares` and `y_shares` fields, 
+for their `nullifier` in an `epoch`.
+The secret can be retreived by the properties of the Shamir's secret sharing scheme. 
+In particular the secret (`a_0`) can be retrieved by computing `Lagrange polynomials`. 
+
+After the secret is retreived,
+ the user's `identity_commitment` can be generated from the secret and it can be used for removing the user from the membership merkle tree (zeroing out the leaf that contains the user's `identity_commitment`). 
+Additionally, depending on the application the `identity_secret_hash` can be used for taking the user's provided stake.
+
+# Auxiliary tooling
```

Comment:
What about making this a subsection under something like "implementation notes"? E.g. see https://rfc.vac.dev/spec/10/#appendix-c-implementation-notes

Then we can mention:

- Circom-based RLN is recommended because of small differences and having interop - also link to this Circuit implementation
- However existing ones like https://github.com/kilic/rln work
- Any other notes that are useful for people who would be *implementing* or analyzing RLN, or *using* it

Reply:
Seems like a great idea